vortex-duckdb/src/convert/array/mod.rs

```diff
@@ -41,53 +67,92 @@ pub fn to_duckdb(array: ArrayRef, chunk: &mut dyn WritableVector) -> VortexResul
     } else if array.is_encoding(DictEncoding.id()) {
         array
             .as_any()
             .downcast_ref::<DictArray>()
             .vortex_expect("dict id checked")
-            .to_duckdb(chunk)
+            .to_duckdb(chunk, cache)
     } else {
-        to_arrow_preferred(&array)?.to_duckdb(chunk)
+        to_arrow_preferred(array)?.to_duckdb(chunk, cache)
     }
 }
 
 impl ToDuckDB for ChunkedArray {
-    fn to_duckdb(&self, chunk: &mut dyn WritableVector) -> VortexResult<()> {
+    fn to_duckdb(
+        &self,
+        chunk: &mut dyn WritableVector,
+        cache: &mut ConversionCache,
+    ) -> VortexResult<()> {
         // TODO(joe): support multi-chunk arrays without canonical.
         if self.chunks().len() > 1 {
-            to_arrow_preferred(self)?.to_duckdb(chunk)
+            to_arrow_preferred(self)?.to_duckdb(chunk, cache)
         } else {
-            to_duckdb(self.chunks()[0].clone(), chunk)
+            to_duckdb(&self.chunks()[0], chunk, cache)
         }
     }
 }
 
 impl ToDuckDB for DictArray {
-    fn to_duckdb(&self, chunk: &mut dyn WritableVector) -> VortexResult<()> {
-        // If the values fit into a single vector, we can efficiently delay the take operation.
-        if self.values().len() <= DUCKDB_STANDARD_VECTOR_SIZE && !self.codes().dtype().is_nullable()
-        {
-            to_duckdb(self.values().clone(), chunk)?;
-            let sel = selection_vector_from_array(self.codes().to_primitive()?);
-            let mut vector = chunk.flat_vector();
-            vector.slice(sel);
-            // Note you can only have nullable values (not codes/selection vectors),
-            // so we cannot assign a selection vector.
-            Ok(())
-        } else {
-            // TODO(joe): can we do value compression and avoid a take.
-            // If the ratio of values to code is low, aka few values to many codes.
+    fn to_duckdb(
+        &self,
+        chunk: &mut dyn WritableVector,
+        cache: &mut ConversionCache,
+    ) -> VortexResult<()> {
+        // Note you can only have nullable values (not codes/selection vectors),
+        // so we cannot assign a selection vector.
+        if !self.codes().all_valid()? {
             let values = take(self.values(), self.codes())?;
-            to_duckdb(values, chunk)
-        }
+            return to_duckdb(&values, chunk, cache);
+        };
+
+        let value_ptr = (self.values().as_ref() as *const dyn Array as *const ()) as usize;
+
+        let mut vector: FlatVector = if self.values().len() <= DUCKDB_STANDARD_VECTOR_SIZE {
+            // If the values fit into a single vector, put the values in the pre-allocated vector.
+            to_duckdb(self.values(), chunk, cache)?;
+            chunk.flat_vector()
+        } else {
+            // If the values don't fit allocated a larger vector and that the data chunk vector
+            // reference this new one.
+            let entry = cache.values_cache.get(&value_ptr);
```

Comment:
There's an API for this pattern:
```
    let value_vector = cache.values_cache
        .entry(value_ptr)
        .or_insert_with(|| {
        //...
```

Reply:
I cannot, since there is a result in the lambda